### ISSUE TITLE
[User] Huge Page Support

### ISSUE THREAD

**Issue by Micraow** [ggerganov/llama.cpp#2251]
Hello, I am using llama.cpp on linux, due to my small memory, my program can often take up 95% of my memory, and I use CPU to run the model, but there's a problem, the `kswapd0` takes about 30% of my CPU. I learned that this process is to change memory page, but it takes too much of my CPU. I wonder if llama.cpp can support huge page and solve this problem.

Thank you

**Comment by ggerganov** [comment]
I'm unfamiliar with the concept of huge pages. If is something that can be achieved with small amount of changes, then we can support it

**Comment by staviq** [comment]
Hugepages is a feature of the Linux kernel, I am not sure if you can programmatically force to use hugepages specifically, but I found this project/library, which can tell you if the memory you are using is/isn't using hugepages, so it might be a good starting point:

https://github.com/travisdowns/page-info

**Comment by slaren** [comment]
I think this would only require adding the `MAP_HUGETLB` flag to `mmap`.

**Comment by qdacsvx** [comment]
MAP_HUGETLB isn't supported for file-backed mmaps. But (on Linux) you can use madvise with MADV_HUGEPAGE.

**Comment by Chris611** [comment]
Any updates on this? For me it would be more for convenience as I've already have most of my memory allocated as hugepages for use with Virtual Machines. And it might be slightly faster

**Comment by jboero** [comment]
I know this is closed but Just a note since I've experimented a bit and tried out the following in every combination while inferencing purely in CPU on 11th gen x86_64 (Tiger Lake) with AVX512 enabled:

1. MADV_HUGEPAGE with transparent hugepages enabled. 
2. Pre-allocated and run hugepages manually.
3. Tried both with using `--mlock` which loads the model into RAM.

In my experience it makes no noticeable performance difference sadly. Since x86 still uses a 4KB pagesize natively I don't think there is much gain to be had with this optimization. Apple Arm currently uses 16KB page size natively which definitely does help when running both Mac OS and Linux (Asahi Fedora). I've not yet tried CPU inference on a PowerPC/Power box which may have benefits with 64K pages.

Obviously all of these are still impractically slow compared to a model properly fit inside the HBM memory of a dedicated GPU.

**Comment by qdacsvx** [comment]
I've had success using MADV_HUGEPAGE - I can measure that hugepages are being used by "grep Huge /proc/meminfo" - it shows >10GB in hugepages when I'm generating using patched llama. It's important to measure performance under load. If you can detect lower TLB pressure that would indicate under sufficient (TLB) load, performance would be expected to be higher.

File system support for hugepages varies (the newest kernels have more support) - EXT4 does support it.

`madvise(addr, len, MADV_MERGEABLE|MADV_HUGEPAGE);`

find out supported sizes of hugepages:
		`ls /sys/kernel/mm/hugepages`
find out default size of huge pages:
		`grep Huge /proc/meminfo`

Enable transparent huge pages on request:
`echo madvise | sudo tee /sys/kernel/mm/transparent_hugepage/enabled`

**Comment by jboero** [comment]
I saw my hugepages grow also but not that much actually.  How much performance gain do you see? Also what type of CPU are you using? I did this on a laptop with just 4 core i7 16GB. It seems CPU bound, not memory bound.

I should reproduce it on my workstation which has more RAM and 44 cores.

**Comment by milpster** [comment]
Trying to get this to work, but it always says that malloc failed to allocate transparent huge pages. I thought it might be a permission thing but even running as root did not change it.

```
impl: mmap with huge page size 2 MiB failed (Cannot allocate memory)
```

```
# cat /proc/meminfo | grep Huge
AnonHugePages:     34816 kB
ShmemHugePages:        0 kB
FileHugePages:     36864 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB

# cat /sys/kernel/mm/transparent_hugepage/enabled 
[always] madvise never
```

**Comment by doctorjei** [comment]
First, hypothetically, this is an easy fix - it's just swapping out the allocation calls (with a check for whether the system supports huge pages). I will try when/if I have some time. (Hopefully soon!)

This problem becomes more "real" as memory gets higher. Machines with 128GiB (like the unified memory architectures that are a part of Strix systems and similar) require ***2 GiB more*** with "normal" pages than with "huge pages". This may not seem like a lot, but it has enormous implications, especially when connected to recent cache strategies (such as "TurboQuant".)

Here's a concrete example on my Strix Halo system with 128GiB of unified RAM.

"Typical" setup:
Total Physical RAM: 131,072 MiB
Explicit VRAM Allocation: 512 MiB <-- minimum (at least on my system)
Firmware Reserved Mem: 230 MiB
Page Array Reservation: 2048 MiB <-- this is what changes
Linux Kernel, with : 790 MiB (approx)
Headless System Services: 150 MiB
Available for llama.cpp: 127,342 MiB (best case scenario, with nothing else running)

"Huge Pages" allows huge pages to be allocated (e.g., 1 GiB) with just an 8 KiB entry, with the remaining 16 MiB released back to available memory. The drawback is that those huge pages are reserved ONLY for huge page requests (and are thus unavailable via "regular" memory calls).

"Huge Pages" setup (if llama.cpp had support):
Total Physical RAM: 131,072 MiB
Reserved (various): 742 MiB (unchanged)
Kernel & System Services: 940 MiB (unchanged)
Page Array Reservation: 1 MiB (If all 128 GiB are allocated as large pages)
Available for llama.cpp: 129,389 MiB (best case scenario)
-----> **Increase of 2047 MiB**

That's a best case scenario, of course. But, here's ubergarm's IQ4_XS for MiniMax m2.5 with tq3 active for KV cache (llama fork):

*Buffers*
Model, CPU: 329.70 MiB
Model, ROCm: 117268.78 MiB
Output, ROCm: 3.05 MiB
KV cache: 9300.13 MiB (for 192Ktoken context size)
Compute, ROCm: 396.75 MiB
Compute, ROCm Host: 212.01 MiB

The ROCme runtime also requires 200-400 MiB, and llama.cpp of course takes up some memory. Before llama.cpp's memory footprint, that's 127910.42 MiB, which is over the "normal" setup, but within the "huge pages" setup.

Basically, those last couple of GiB matter practically - it's the difference between a robust context size and a much smaller one.
